File: src/tauri/commands/campaigns.ts
import { invoke } from "@tauri-apps/api/core";
import {
  Campaign,
  CampaignSummary,
  CreateCampaignRequest,
  UpdateCampaignRequest,
  AppSettings,
} from "@/core/entities/Campaign";

/**
 * Typed wrapper for Tauri campaign commands
 * Provides type safety and error handling for campaign operations
 */
export class CampaignApi {
  /**
   * Create a new campaign
   */
  static async create(request: CreateCampaignRequest): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("create_campaign", {
        req: request,
      });
      return result;
    } catch (error) {
      console.error("Failed to create campaign:", error);
      throw new Error(`Failed to create campaign: ${error}`);
    }
  }

  /**
   * Get campaign by ID
   */
  static async getById(campaignId: string): Promise<Campaign | null> {
    try {
      const result = await invoke<Campaign | null>("get_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to get campaign:", error);
      throw new Error(`Failed to get campaign: ${error}`);
    }
  }

  /**
   * Get all campaigns
   */
  static async getAll(): Promise<Campaign[]> {
    try {
      const result = await invoke<Campaign[]>("get_all_campaigns");
      return result;
    } catch (error) {
      console.error("Failed to get all campaigns:", error);
      throw new Error(`Failed to get all campaigns: ${error}`);
    }
  }

  /**
   * Get only active campaigns
   */
  static async getActive(): Promise<Campaign[]> {
    try {
      const result = await invoke<Campaign[]>("get_active_campaigns");
      return result;
    } catch (error) {
      console.error("Failed to get active campaigns:", error);
      throw new Error(`Failed to get active campaigns: ${error}`);
    }
  }

  /**
   * Get campaign summaries for list views
   */
  static async getSummaries(): Promise<CampaignSummary[]> {
    try {
      const result = await invoke<CampaignSummary[]>("get_campaign_summaries");
      return result;
    } catch (error) {
      console.error("Failed to get campaign summaries:", error);
      throw new Error(`Failed to get campaign summaries: ${error}`);
    }
  }

  /**
   * Update campaign data
   */
  static async update(
    campaignId: string,
    request: UpdateCampaignRequest
  ): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("update_campaign", {
        campaignId,
        req: request,
      });
      return result;
    } catch (error) {
      console.error("Failed to update campaign:", error);
      throw new Error(`Failed to update campaign: ${error}`);
    }
  }

  /**
   * Delete campaign
   */
  static async delete(campaignId: string): Promise<boolean> {
    try {
      const result = await invoke<boolean>("delete_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to delete campaign:", error);
      throw new Error(`Failed to delete campaign: ${error}`);
    }
  }

  /**
   * Archive campaign (safer than delete)
   */
  static async archive(campaignId: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("archive_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to archive campaign:", error);
      throw new Error(`Failed to archive campaign: ${error}`);
    }
  }

  /**
   * Set current active campaign
   */
  static async setCurrent(campaignId: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("set_current_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to set current campaign:", error);
      throw new Error(`Failed to set current campaign: ${error}`);
    }
  }

  /**
   * Get current active campaign
   */
  static async getCurrent(): Promise<Campaign | null> {
    try {
      const result = await invoke<Campaign | null>("get_current_campaign");
      return result;
    } catch (error) {
      console.error("Failed to get current campaign:", error);
      throw new Error(`Failed to get current campaign: ${error}`);
    }
  }

  /**
   * Clear current campaign
   */
  static async clearCurrent(): Promise<void> {
    try {
      await invoke<void>("clear_current_campaign");
    } catch (error) {
      console.error("Failed to clear current campaign:", error);
      throw new Error(`Failed to clear current campaign: ${error}`);
    }
  }

  /**
   * Get recently accessed campaigns
   */
  static async getRecent(): Promise<Campaign[]> {
    try {
      const result = await invoke<Campaign[]>("get_recent_campaigns");
      return result;
    } catch (error) {
      console.error("Failed to get recent campaigns:", error);
      throw new Error(`Failed to get recent campaigns: ${error}`);
    }
  }

  /**
   * Start a new session in campaign
   */
  static async startSession(campaignId: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("start_campaign_session", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to start campaign session:", error);
      throw new Error(`Failed to start campaign session: ${error}`);
    }
  }

  /**
   * Update campaign statistics
   */
  static async updateStats(
    campaignId: string,
    activeCharacters: number,
    totalCharacters: number,
    averageLevel: number
  ): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("update_campaign_stats", {
        campaignId,
        activeCharacters,
        totalCharacters,
        averageLevel,
      });
      return result;
    } catch (error) {
      console.error("Failed to update campaign stats:", error);
      throw new Error(`Failed to update campaign stats: ${error}`);
    }
  }

  /**
   * Complete campaign
   */
  static async complete(campaignId: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("complete_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to complete campaign:", error);
      throw new Error(`Failed to complete campaign: ${error}`);
    }
  }

  /**
   * Pause campaign
   */
  static async pause(campaignId: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("pause_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to pause campaign:", error);
      throw new Error(`Failed to pause campaign: ${error}`);
    }
  }

  /**
   * Resume campaign
   */
  static async resume(campaignId: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("resume_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to resume campaign:", error);
      throw new Error(`Failed to resume campaign: ${error}`);
    }
  }

  /**
   * Duplicate campaign with new name
   */
  static async duplicate(
    campaignId: string,
    newName: string
  ): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("duplicate_campaign", {
        campaignId,
        newName,
      });
      return result;
    } catch (error) {
      console.error("Failed to duplicate campaign:", error);
      throw new Error(`Failed to duplicate campaign: ${error}`);
    }
  }

  /**
   * Export campaign to JSON
   */
  static async export(campaignId: string): Promise<string> {
    try {
      const result = await invoke<string>("export_campaign", {
        campaignId,
      });
      return result;
    } catch (error) {
      console.error("Failed to export campaign:", error);
      throw new Error(`Failed to export campaign: ${error}`);
    }
  }

  /**
   * Import campaign from JSON
   */
  static async import(jsonData: string): Promise<Campaign> {
    try {
      const result = await invoke<Campaign>("import_campaign", {
        jsonData,
      });
      return result;
    } catch (error) {
      console.error("Failed to import campaign:", error);
      throw new Error(`Failed to import campaign: ${error}`);
    }
  }

  /**
   * Backup all campaigns
   */
  static async backup(): Promise<string> {
    try {
      const result = await invoke<string>("backup_campaigns");
      return result;
    } catch (error) {
      console.error("Failed to backup campaigns:", error);
      throw new Error(`Failed to backup campaigns: ${error}`);
    }
  }

  /**
   * Get app settings
   */
  static async getSettings(): Promise<AppSettings> {
    try {
      const result = await invoke<AppSettings>("get_app_settings");
      return result;
    } catch (error) {
      console.error("Failed to get app settings:", error);
      throw new Error(`Failed to get app settings: ${error}`);
    }
  }

  /**
   * Update app theme
   */
  static async updateTheme(theme: string): Promise<AppSettings> {
    try {
      const result = await invoke<AppSettings>("update_app_theme", {
        theme,
      });
      return result;
    } catch (error) {
      console.error("Failed to update theme:", error);
      throw new Error(`Failed to update theme: ${error}`);
    }
  }

  /**
   * Update backup settings
   */
  static async updateBackupSettings(
    autoBackup: boolean,
    backupFrequencyHours: number
  ): Promise<AppSettings> {
    try {
      const result = await invoke<AppSettings>("update_backup_settings", {
        autoBackup,
        backupFrequencyHours,
      });
      return result;
    } catch (error) {
      console.error("Failed to update backup settings:", error);
      throw new Error(`Failed to update backup settings: ${error}`);
    }
  }

  /**
   * Validate campaign name for uniqueness
   */
  static async validateName(name: string): Promise<boolean> {
    try {
      const result = await invoke<boolean>("validate_campaign_name", {
        name,
      });
      return result;
    } catch (error) {
      console.error("Failed to validate campaign name:", error);
      throw new Error(`Failed to validate campaign name: ${error}`);
    }
  }

  /**
   * Get campaign count
   */
  static async getCount(): Promise<number> {
    try {
      const result = await invoke<number>("get_campaign_count");
      return result;
    } catch (error) {
      console.error("Failed to get campaign count:", error);
      throw new Error(`Failed to get campaign count: ${error}`);
    }
  }

  /**
   * Get active campaign count
   */
  static async getActiveCount(): Promise<number> {
    try {
      const result = await invoke<number>("get_active_campaign_count");
      return result;
    } catch (error) {
      console.error("Failed to get active campaign count:", error);
      throw new Error(`Failed to get active campaign count: ${error}`);
    }
  }

  /**
   * Check if campaigns file exists
   */
  static async fileExists(): Promise<boolean> {
    try {
      const result = await invoke<boolean>("campaigns_file_exists");
      return result;
    } catch (error) {
      console.error("Failed to check if campaigns file exists:", error);
      throw new Error(`Failed to check if campaigns file exists: ${error}`);
    }
  }

  /**
   * Initialize app (create default files if needed)
   */
  static async initializeApp(): Promise<string> {
    try {
      const result = await invoke<string>("initialize_app");
      return result;
    } catch (error) {
      console.error("Failed to initialize app:", error);
      throw new Error(`Failed to initialize app: ${error}`);
    }
  }
}

// Convenience functions for common operations
export const campaignsApi = {
  // CRUD operations
  create: CampaignApi.create,
  getById: CampaignApi.getById,
  getAll: CampaignApi.getAll,
  getActive: CampaignApi.getActive,
  getSummaries: CampaignApi.getSummaries,
  update: CampaignApi.update,
  delete: CampaignApi.delete,
  archive: CampaignApi.archive,

  // Current campaign management
  setCurrent: CampaignApi.setCurrent,
  getCurrent: CampaignApi.getCurrent,
  clearCurrent: CampaignApi.clearCurrent,
  getRecent: CampaignApi.getRecent,

  // Session management
  startSession: CampaignApi.startSession,
  updateStats: CampaignApi.updateStats,

  // Campaign lifecycle
  complete: CampaignApi.complete,
  pause: CampaignApi.pause,
  resume: CampaignApi.resume,
  duplicate: CampaignApi.duplicate,

  // Import/Export
  export: CampaignApi.export,
  import: CampaignApi.import,
  backup: CampaignApi.backup,

  // Settings
  getSettings: CampaignApi.getSettings,
  updateTheme: CampaignApi.updateTheme,
  updateBackupSettings: CampaignApi.updateBackupSettings,

  // Utilities
  validateName: CampaignApi.validateName,
  getCount: CampaignApi.getCount,
  getActiveCount: CampaignApi.getActiveCount,
  fileExists: CampaignApi.fileExists,
  initializeApp: CampaignApi.initializeApp,
};

export default campaignsApi;

File: src-tauri/src/commands/campaigns.rs
use crate::core::models::campaign::{
    Campaign, 
    CampaignSummary,
    CreateCampaignRequest, 
    UpdateCampaignRequest,
    AppSettings,
    CampaignStatus,
};
use crate::services::campaign_storage::CampaignStorageService;
use crate::utils::error::{DmAssistantError, DmResult, validate_uuid, validate_non_empty};
use uuid::Uuid;

// ============================================
// CORE CAMPAIGN OPERATIONS
// ============================================

#[tauri::command]
pub async fn create_campaign(req: CreateCampaignRequest) -> DmResult<Campaign> {
    // Validate required fields
    validate_non_empty(&req.name, "Campaign name")?;
    validate_non_empty(&req.description, "Campaign description")?;
    validate_non_empty(&req.setting, "Campaign setting")?;
    
    // Validate player count if provided
    if let Some(count) = req.player_count {
        if count == 0 || count > 12 {
            return Err(DmAssistantError::validation("Player count must be between 1 and 12"));
        }
    }
    
    let campaign = Campaign::new(req);
    CampaignStorageService::create_campaign(campaign)
}

#[tauri::command]
pub async fn get_campaign(campaign_id: String) -> DmResult<Option<Campaign>> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    CampaignStorageService::get_campaign(&id)
}

#[tauri::command]
pub async fn get_all_campaigns() -> DmResult<Vec<Campaign>> {
    CampaignStorageService::get_all_campaigns()
}

#[tauri::command]
pub async fn get_active_campaigns() -> DmResult<Vec<Campaign>> {
    CampaignStorageService::get_active_campaigns()
}

#[tauri::command]
pub async fn get_campaign_summaries() -> DmResult<Vec<CampaignSummary>> {
    CampaignStorageService::get_campaign_summaries()
}

#[tauri::command]
pub async fn update_campaign(
    campaign_id: String,
    req: UpdateCampaignRequest,
) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    
    // Validate fields if provided
    if let Some(ref name) = req.name {
        validate_non_empty(name, "Campaign name")?;
    }
    if let Some(ref description) = req.description {
        validate_non_empty(description, "Campaign description")?;
    }
    if let Some(ref setting) = req.setting {
        validate_non_empty(setting, "Campaign setting")?;
    }
    if let Some(count) = req.player_count {
        if count == 0 || count > 12 {
            return Err(DmAssistantError::validation("Player count must be between 1 and 12"));
        }
    }
    
    CampaignStorageService::modify_campaign(&id, |campaign| {
        campaign.update(req);
        Ok(())
    })
}

#[tauri::command]
pub async fn delete_campaign(campaign_id: String) -> DmResult<bool> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    CampaignStorageService::delete_campaign(&id)
}

#[tauri::command]
pub async fn archive_campaign(campaign_id: String) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    CampaignStorageService::archive_campaign(&id)
}

// ============================================
// CURRENT CAMPAIGN MANAGEMENT
// ============================================

#[tauri::command]
pub async fn set_current_campaign(campaign_id: String) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    CampaignStorageService::set_current_campaign(&id)
}

#[tauri::command]
pub async fn get_current_campaign() -> DmResult<Option<Campaign>> {
    CampaignStorageService::get_current_campaign()
}

#[tauri::command]
pub async fn clear_current_campaign() -> DmResult<()> {
    CampaignStorageService::clear_current_campaign()
}

#[tauri::command]
pub async fn get_recent_campaigns() -> DmResult<Vec<Campaign>> {
    CampaignStorageService::get_recent_campaigns()
}

// ============================================
// SESSION MANAGEMENT
// ============================================

#[tauri::command]
pub async fn start_campaign_session(campaign_id: String) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    CampaignStorageService::start_session(&id)
}

#[tauri::command]
pub async fn update_campaign_stats(
    campaign_id: String,
    active_characters: u32,
    total_characters: u32,
    average_level: f32,
) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    
    // Validate stats
    if total_characters < active_characters {
        return Err(DmAssistantError::validation("Total characters cannot be less than active characters"));
    }
    if average_level < 1.0 || average_level > 20.0 {
        return Err(DmAssistantError::validation("Average level must be between 1.0 and 20.0"));
    }
    
    CampaignStorageService::update_campaign_stats(&id, active_characters, total_characters, average_level)
}

// ============================================
// CAMPAIGN LIFECYCLE
// ============================================

#[tauri::command]
pub async fn complete_campaign(campaign_id: String) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    
    CampaignStorageService::modify_campaign(&id, |campaign| {
        if !campaign.is_playable() {
            return Err(DmAssistantError::campaign("Cannot complete a non-playable campaign"));
        }
        campaign.complete();
        Ok(())
    })
}

#[tauri::command]
pub async fn pause_campaign(campaign_id: String) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    
    CampaignStorageService::modify_campaign(&id, |campaign| {
        if !matches!(campaign.campaign_info.campaign_status, CampaignStatus::Active) {
            return Err(DmAssistantError::campaign("Can only pause active campaigns"));
        }
        campaign.campaign_info.campaign_status = CampaignStatus::OnHold;
        Ok(())
    })
}

#[tauri::command]
pub async fn resume_campaign(campaign_id: String) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    
    CampaignStorageService::modify_campaign(&id, |campaign| {
        if !matches!(campaign.campaign_info.campaign_status, CampaignStatus::OnHold) {
            return Err(DmAssistantError::campaign("Can only resume paused campaigns"));
        }
        campaign.campaign_info.campaign_status = CampaignStatus::Active;
        Ok(())
    })
}

#[tauri::command]
pub async fn duplicate_campaign(
    campaign_id: String, 
    new_name: String
) -> DmResult<Campaign> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    validate_non_empty(&new_name, "New campaign name")?;
    
    // Get original campaign
    let original = CampaignStorageService::get_campaign(&id)?
        .ok_or_else(|| DmAssistantError::not_found("Campaign", &campaign_id))?;
    
    // Create duplicate with new ID and name
    let mut duplicate = original.clone();
    duplicate.id = Uuid::new_v4();
    duplicate.name = new_name;
    duplicate.current_session = 0;
    duplicate.campaign_info.total_sessions = 0;
    duplicate.campaign_info.campaign_status = CampaignStatus::Planning;
    duplicate.last_session_date = None;
    duplicate.created_at = chrono::Utc::now();
    duplicate.updated_at = chrono::Utc::now();
    
    // Reset character-related stats (will be populated when characters are added)
    duplicate.campaign_info.total_characters = 0;
    duplicate.campaign_info.active_characters = 0;
    duplicate.player_count = 0;
    duplicate.average_level = 1.0;
    
    CampaignStorageService::create_campaign(duplicate)
}

// ============================================
// IMPORT/EXPORT & BACKUP
// ============================================

#[tauri::command]
pub async fn export_campaign(campaign_id: String) -> DmResult<String> {
    let id = validate_uuid(&campaign_id, "Campaign")?;
    CampaignStorageService::export_campaign(&id)
}

#[tauri::command]
pub async fn import_campaign(json_data: String) -> DmResult<Campaign> {
    validate_non_empty(&json_data, "JSON data")?;
    CampaignStorageService::import_campaign(&json_data)
}

#[tauri::command]
pub async fn backup_campaigns() -> DmResult<String> {
    let backup_path = CampaignStorageService::backup_campaigns()?;
    Ok(format!("Campaigns backed up successfully to: {}", backup_path.display()))
}

// ============================================
// APP SETTINGS
// ============================================

#[tauri::command]
pub async fn get_app_settings() -> DmResult<AppSettings> {
    CampaignStorageService::get_app_settings()
}

#[tauri::command]
pub async fn update_app_theme(theme: String) -> DmResult<AppSettings> {
    validate_non_empty(&theme, "Theme")?;
    
    // Validate theme value
    if !["light", "dark", "system"].contains(&theme.as_str()) {
        return Err(DmAssistantError::validation("Theme must be 'light', 'dark', or 'system'"));
    }
    
    CampaignStorageService::update_app_settings(|settings| {
        settings.theme = theme;
        Ok(())
    })
}

#[tauri::command]
pub async fn update_backup_settings(
    auto_backup: bool,
    backup_frequency_hours: u32,
) -> DmResult<AppSettings> {
    // Validate backup frequency
    if backup_frequency_hours == 0 || backup_frequency_hours > 168 { // Max 1 week
        return Err(DmAssistantError::validation("Backup frequency must be between 1 and 168 hours"));
    }
    
    CampaignStorageService::update_app_settings(|settings| {
        settings.auto_backup = auto_backup;
        settings.backup_frequency_hours = backup_frequency_hours;
        Ok(())
    })
}

// ============================================
// UTILITY FUNCTIONS
// ============================================

#[tauri::command]
pub async fn validate_campaign_name(name: String) -> DmResult<bool> {
    validate_non_empty(&name, "Campaign name")?;
    
    // Check if name already exists
    let campaigns = CampaignStorageService::get_all_campaigns()?;
    let exists = campaigns.iter().any(|c| c.name.eq_ignore_ascii_case(&name));
    
    Ok(!exists) // Return true if name is available (doesn't exist)
}

#[tauri::command]
pub async fn get_campaign_count() -> DmResult<u32> {
    let campaigns = CampaignStorageService::get_all_campaigns()?;
    Ok(campaigns.len() as u32)
}

#[tauri::command]
pub async fn get_active_campaign_count() -> DmResult<u32> {
    let campaigns = CampaignStorageService::get_active_campaigns()?;
    Ok(campaigns.len() as u32)
}

#[tauri::command]
pub async fn campaigns_file_exists() -> DmResult<bool> {
    Ok(CampaignStorageService::campaigns_file_exists())
}

#[tauri::command]
pub async fn initialize_app() -> DmResult<String> {
    CampaignStorageService::initialize_app()?;
    
    // Return status message
    let campaigns_exist = CampaignStorageService::campaigns_file_exists();
    let app_settings = CampaignStorageService::get_app_settings()?;
    let current_campaign = CampaignStorageService::get_current_campaign()?;
    
    let status = if campaigns_exist {
        format!(
            "App initialized successfully. Current campaign: {}",
            current_campaign
                .map(|c| c.name)
                .unwrap_or_else(|| "None".to_string())
        )
    } else {
        "App initialized with default settings. No campaigns found.".to_string()
    };
    
    Ok(status)
}

// ============================================
// HELPER FUNCTIONS FOR VALIDATION
// ============================================

/// Validate that campaign exists and return it
async fn get_campaign_or_error(campaign_id: &str) -> DmResult<Campaign> {
    let id = validate_uuid(campaign_id, "Campaign")?;
    CampaignStorageService::get_campaign(&id)?
        .ok_or_else(|| DmAssistantError::not_found("Campaign", campaign_id))
}

/// Validate that campaign is in a state that allows modifications
fn validate_campaign_modifiable(campaign: &Campaign) -> DmResult<()> {
    if !campaign.campaign_info.campaign_status.allows_modifications() {
        return Err(DmAssistantError::campaign(
            "Cannot modify campaign in current status. Only Planning, Active, and OnHold campaigns can be modified."
        ));
    }
    Ok(())
}

/// Validate campaign name uniqueness (excluding specific campaign)
async fn validate_name_unique_excluding(name: &str, exclude_id: &Uuid) -> DmResult<()> {
    let campaigns = CampaignStorageService::get_all_campaigns()?;
    let exists = campaigns.iter().any(|c| c.id != *exclude_id && c.name.eq_ignore_ascii_case(name));
    
    if exists {
        return Err(DmAssistantError::validation("Campaign name already exists"));
    }
    
    Ok(())
}

// ============================================
// ADVANCED OPERATIONS
// ============================================

#[tauri::command]
pub async fn get_campaign_analytics() -> DmResult<CampaignAnalytics> {
    let campaigns = CampaignStorageService::get_all_campaigns()?;
    
    let total_campaigns = campaigns.len() as u32;
    let active_campaigns = campaigns.iter().filter(|c| c.is_active).count() as u32;
    let completed_campaigns = campaigns.iter()
        .filter(|c| matches!(c.campaign_info.campaign_status, CampaignStatus::Completed))
        .count() as u32;
    
    let total_sessions: u32 = campaigns.iter().map(|c| c.campaign_info.total_sessions).sum();
    let total_characters: u32 = campaigns.iter().map(|c| c.campaign_info.total_characters).sum();
    
    let avg_level = if total_characters > 0 {
        campaigns.iter()
            .map(|c| c.average_level * c.campaign_info.active_characters as f32)
            .sum::<f32>() / total_characters as f32
    } else {
        0.0
    };
    
    Ok(CampaignAnalytics {
        total_campaigns,
        active_campaigns,
        completed_campaigns,
        total_sessions,
        total_characters,
        average_level: avg_level,
    })
}

/// Analytics data for dashboard
#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CampaignAnalytics {
    pub total_campaigns: u32,
    pub active_campaigns: u32,
    pub completed_campaigns: u32,
    pub total_sessions: u32,
    pub total_characters: u32,
    pub average_level: f32,
}